lib/puppet/provider/user/windows_adsi.rb

```diff
@@ -123,8 +306,58 @@ def password
     user.password_is?( @resource[:password] ) ? @resource[:password] : nil
   end
 
+  # Sets the user's password. Takes the following options:
+  #
+  #   * force -- If true, the user's password will be set
+  #              no matter what.
+  #
+  def set_password(password, options = {})
+    force = options[:force] || false
+
+    # The force option is useful here because password_change_required will
+    # always be true after a user's created. Thus without the force option,
+    # we cannot set the user's password after creating them.
+    current_attributes = attributes
+    if ! force && (current_attributes[:password_change_required] == 'true')
+      warning _("The account '%{name}' requires the user to change their password on the next sign-in. Refusing to reset the password. If you still wish to reset the user's password, then set password_change_required to false in the attributes property and then re-run Puppet.") % { name: name }
+      return
+    end
+
+    user.password = password
+
+    if (attributes_prop = @resource.property(:attributes))
+      specified_attributes = attributes_prop.hashify_should
+      unless specified_attributes.key?(:password_never_expires)
+        warning _("The attributes property did not specify the 'password_never_expires' attribute, which is currently set to %{password_never_expires}. If this is undesirable behavior, then please specify a value for the 'password_never_expires' attribute, then re-run Puppet.") % { password_never_expires: current_attributes[:password_never_expires] }
```

Comment:
I figured this warning message would be useful to people who would expect the `password_never_expires` attribute to be set by default.